File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.ts
function firstStableIndex(nums: number[], k: number): number {
    let maxSoFar = -Infinity;
    let cand = 0;
    let cMax = 0;
    const n = nums.length;

    for (let i = 0; i < n; i++) {
        maxSoFar = Math.max(maxSoFar, nums[i]);

        if (i === cand) cMax = maxSoFar;

        if (nums[i] < cMax - k) {
            cand = i + 1;
        }
    }

    return cand < n ? cand : -1;
};

File: 3904-smallest-stable-index-ii/3904-smallest-stable-index-ii.rs
impl Solution {
    pub fn first_stable_index(nums: Vec<i32>, k: i32) -> i32 {
        let mut max_so_far = i32::MIN;
        let mut cand = 0;
        let mut c_max = 0;
        let n = nums.len();

        for i in 0..n {
            max_so_far = max_so_far.max(nums[i]);

            if i == cand {
                c_max = max_so_far;
            }

            if nums[i] < c_max - k {
                cand = i + 1;
            }
        }

        if cand < n { cand as i32 } else { -1 }
    }
}